### ISSUE TITLE
mingw-w64-qt5-base (5.7.0-8): CMake qt5_wrap_ui adds Qt5::uic to uic command line

### ISSUE THREAD

**Issue by JayFoxRox** [Martchus/PKGBUILDs#11]
I'm trying to compile github.com/citra-emu/citra with mingw.

I'm getting the following problem (mingw-w64-qt5-base 5.7.0-8 on both x64 and i686 using the mingw CMake wrappers):

```
[ 86%] Generating ui_main.h
cd /home/fox/Data/Projects/citra/build-x86_64-w64-mingw32/src/citra_qt && /usr/x86_64-w64-mingw32/lib/qt/bin/uic Qt5::uic -o /home/fox/Data/Projects/citra/build-x86_64-w64-mingw32/src/citra_qt/ui_main.h /home/fox/Data/Projects/citra/src/citra_qt/main.ui
File 'Qt5::uic' is not valid
```

A native Linux build (qt5-base 5.7.0-7 x64) does this:

```
[ 87%] Generating ui_main.h
cd /home/fox/Data/Projects/citra/build/src/citra_qt && /usr/bin/uic -o /home/fox/Data/Projects/citra/build/src/citra_qt/ui_main.h /home/fox/Data/Projects/citra/src/citra_qt/main.ui
```

It appears that CMake `qt5_wrap_ui(UI_HDRS ${UIS})` (or some wrapper invoking `uic`) provided by the mingw qt5 is somehow broken?
Unfortunately the citra codebase is not AUTOUIC ready, so I did not try it with that yet.

---

As this issue does not exist natively I'm assuming it's a bug in the package?
Is there anything else you can recommend I should try or is there a known workaround for this issue?

I've tried finding other people who are affected but found nothing yet.

**Comment by Martchus** [comment]
Yes, looks like a bug in the package or the mingw-w64/Windows version of Qt. Since I lately patched the CMake files in order to provide support to use static build with CMake I might have messed something.

The latest version (5.7.0-9, found in the repository on GitHub and my binary repo) might work better, but I wouldn't bet on it.

AUTOUIC would work in general. I use it in my own projects instead of `qt5_wrap_ui` so I didn't notice that this function is broken.

I'll look into this next week. Until then, you could just *not* apply the patch `0022-Allow-usage-of-static-version-with-CMake.patch` (which is the mentioned patch for static build). Other patches also modify the CMake files (like `0020-Use-.dll.a-as-import-lib-extension.patch`) and hance are also possible candidates.

You can also play with the installed CMake files. Maybe you find a fix. I can include those changes into the package then. Just in case you can provide a patch, patches are managed [here](https://github.com/Martchus/qtbase).

**Comment by JayFoxRox** [comment]
I've tried to find the cause of this and I'm absolutely clueless what goes on.
Aside from a single change in /usr/i686-w64-mingw32/lib/cmake/Qt5Widgets/Qt5WidgetsConfigExtras.cmake (prefixing the path) the (Qt5Widgets) CMake files seem to be the same for linux, windows and mingw.

---

I noticed the following behaviour in /usr/i686-w64-mingw32/lib/cmake/Qt5Widgets/Qt5WidgetsMacros.cmake . I'm not sure if that means anything though:

```
add_custom_command(OUTPUT ${outfile}
          COMMAND ${Qt5Widgets_UIC_EXECUTABLE}
```

Will expand ${Qt5Widgets_UIC_EXECUTABLE} to "/usr/i686-w64-mingw32/lib/qt/bin/uic Qt5::uic"

```
add_custom_command(OUTPUT ${outfile}
          COMMAND FOO ${Qt5Widgets_UIC_EXECUTABLE}
```

Will expand ${Qt5Widgets_UIC_EXECUTABLE} to "FOO Qt5::uic". This suggests that Qt5::uic is supposed to be replaced by the uic path, which doesn't happen, as it's merely prefixed for some strange reason?!

```
add_custom_command(OUTPUT ${outfile}
          COMMAND FOO ${Qt5Widgets_UIC_EXECUTABLE} BAR
```
Will expand ${Qt5Widgets_UIC_EXECUTABLE} to "FOO Qt5::uic BAR"


```
add_custom_command(OUTPUT ${outfile}
          COMMAND ${Qt5Widgets_UIC_EXECUTABLE} BAR
```
Will expand ${Qt5Widgets_UIC_EXECUTABLE} to "/usr/i686-w64-mingw32/lib/qt/bin/uic Qt5::uic BAR"


---

As a temporary workaround I'll attempt to switch to AUTOUIC, maybe compile without 0022 and 0020

**Comment by Martchus** [comment]
Ok, then I guess my patches are not the source of the problem.

But this reminds me of a similar problem with the macros provided by Qt Linguist (part of qt5-tools): https://github.com/Martchus/qttools/commit/1f1a8e6109b99b797efe52e91c47c1c04c629b83

As you can see I've already found a workaround for this. You might be able to apply the same workaround for uic.

**Comment by Martchus** [comment]
If this is correct, the following patch might fix the issue: https://github.com/Martchus/qtbase/commit/7e20d462a43054bc5ff6776454277344436df710

**Comment by Martchus** [comment]
So, does this little change work for you? You don't need to rebuild the package to test, just modify installed file.

**Comment by JayFoxRox** [comment]
Oh I forgot to get back to this. I've implemented my own hack which was a lot more cruel:
```
set(Qt5Widgets_UIC_EXECUTABLE ${_qt5Widgets_install_prefix}/lib/qt/bin/uic)
```

( in /usr/{i686,x86_64}-w64-mingw32/lib/cmake/Qt5Widgets/Qt5WidgetsConfigExtras.cmake )
I only really needed this to test one PR in citra so I didn't try your solution either.

---

However, being well aware how annoying it is to reproduce issues, I've also tried your hack (I wouldn't dare to call it a patch) now: It does also work.
Let's hope you find the real cause of the issue soon so these ugly kludges can be removed.
